### PR TITLE
Add integration for tensor statistic collection via NVIDIA-DLFW-Inspect

### PR DESCRIPTION
**Pull request by paul-gibbons**
### Tensor Inspection (nvidia-dlfw-inspect) integration in Megatron-Bridge via TransformerEngine debug modules

This MR adds support for [NVIDIA DL Framework Inspect](https://github.com/NVIDIA/nvidia-dlfw-inspect/tree/main) to Megatron-Bridge. When enabled, you can collect tensor statistics from TransformerEngine linear modules during training to help debug convergence behavior in BF16 as well as low precision training recipes.

---

### What is nvidia-dlfw-inspect?

**nvidia-dlfw-inspect** is tool that facilitates debugging convergence issues and testing new algorithms/recipes for training LLMs using Nvidia libraries: Transformer Engine, Megatron-LM, and NeMo.

Users can leverage configuration files to enable or disable specific features by selecting layers based on their names. This allows for a more tailored debugging experience, as users can add framework-specific features and only use the features that are necessary, without having to load all debug features.

Additionally, APIs are provided for processing tensors and collecting tensor statistics. These APIs can be easily customized based on user specifications, providing a flexible and adaptable debugging solution.

### Additional Resources 
See NVIDIA-DLFW-Inspect's [documentation](https://github.com/NVIDIA/nvidia-dlfw-inspect/tree/main?tab=readme-ov-file#nvidia-dl-framework-inspect) for further details.

See TransformerEngine's  [integration](https://github.com/NVIDIA/TransformerEngine/tree/fd2f589f26dfedeb4b3705f0137ec89fe8e5d8cf/transformer_engine/debug) & [documentation](https://github.com/NVIDIA/TransformerEngine/tree/fd2f589f26dfedeb4b3705f0137ec89fe8e5d8cf/docs/debug) for comprehensive documentation of how the NVIDIA-DLFW-Inspect tool is integrated in TE's modules.

### CODE REVIEW

File: pyproject.toml [hunk 1]
```diff
@@ -79,6 +79,7 @@ dependencies = [
     "tqdm>=4.67.1",
     "hydra-core>1.3,<=1.3.2",
     "megatron-core[dev,mlm]>=0.14.0a0,<0.16.0",
+    "nvdlfw-inspect==0.2.1"
```
Comment (ananthsub) [inline pyproject.toml]: could this be an optional dependency to start with?

File: src/megatron/bridge/training/tensor_inspect.py [hunk 2]
```diff
@@ -0,0 +1,142 @@
...
+        BaseLogger,
+        MetricLogger,
+        wrap_tensorboard_writer,
+    )
+    HAVE_NVINSPECT = True
+except (ImportError, ModuleNotFoundError):
+    HAVE_NVINSPECT = False
+
+
+def initialize_tensor_inspect_pre_model(cfg: Any, state: Any) -> None:
```
Comment (ananthsub) [inline src/megatron/bridge/training/tensor_inspect.py]: prefer to accept just the configs that are needed. this could accept just the tensor inspect config dataclass

File: src/megatron/bridge/training/tensor_inspect.py [hunk 3]
```diff
@@ -0,0 +1,142 @@
...
+
+    if cfg.tensor_inspect is None or not cfg.tensor_inspect.enabled:
+        return
+
+    if not HAVE_NVINSPECT:
+        print_rank_0(MISSING_NVINSPECT_MSG)
+        raise ImportError(MISSING_NVINSPECT_MSG)
+
+    try:
+        log_dir = cfg.tensor_inspect.log_dir or cfg.checkpoint.save or "."
```
Comment (ananthsub) [inline src/megatron/bridge/training/tensor_inspect.py]: this fallback could be set in the config container validate method to handle the cross-config dependencies. that way only the tensor_inspect dataclass is needed as input here

File: src/megatron/bridge/training/tensor_inspect.py [hunk 4]
```diff
@@ -0,0 +1,142 @@
...
+        nvinspect_api.initialize(
+            config_file=cfg.tensor_inspect.features or "",
+            feature_dirs=cfg.tensor_inspect.feature_dirs,
+            log_dir=log_dir,
+            statistics_logger=None,
+            init_training_step=state.train_state.step,
```
Comment (ananthsub) [inline src/megatron/bridge/training/tensor_inspect.py]: this would always be 0 since this is called before the checkpoint load occurs. is that expected?

File: pyproject.toml (hunk 1 above)
Comment (paul-gibbons) [reply pyproject.toml]: Yes. Should I move to [project.optional-dependencies] under nemo-run?

File: src/megatron/bridge/training/tensor_inspect.py (hunk 4 above)
Comment (paul-gibbons) [reply src/megatron/bridge/training/tensor_inspect.py]: Yes it would always be zero in that case. I will add an additional nvinspect_api call that is run after checkpoint load so it's internal step count will be set properly.

File: src/megatron/bridge/training/tensor_inspect.py (hunk 3 above)
Comment (paul-gibbons) [reply src/megatron/bridge/training/tensor_inspect.py]: ack, moving it to ConfigContainer.

File: src/megatron/bridge/training/tensor_inspect.py (hunk 2 above)
Comment (paul-gibbons) [reply src/megatron/bridge/training/tensor_inspect.py]: Will adjust in next commit.

File: src/megatron/bridge/training/tensor_inspect.py (hunk 4 above)
Comment (anmolgupt) [reply src/megatron/bridge/training/tensor_inspect.py]: You can use the api: https://github.com/NVIDIA/nvidia-dlfw-inspect/blob/main/nvdlfw_inspect/api.py#L90-L93 to update the training step

File: src/megatron/bridge/training/config.py [hunk 5]
```diff
@@ -630,6 +630,45 @@ def __post_init__(self) -> None:
...
+    """Directories containing feature implementations (searched recursively)."""
+
+    log_dir: Optional[str] = None
+    """Root directory to store inspection logs/statistics. Defaults to checkpoint save dir if unset."""
+
+    init_training_step: int = 0
+    """Initial training step for the inspector (used when resuming)."""
+
+
+    def __post_init__(self) -> None:
```
Comment (ananthsub) [inline src/megatron/bridge/training/config.py]: in https://github.com/NVIDIA-NeMo/Megatron-Bridge/pull/627, we've updated the dataclass configs to avoid using `__post_init__`. this better allows for overrides, followed by any deferred initialization to happen once at the beginning of training without risk of dependent attributes not being reset correctly

```suggestion
    def finalize(self) -> None:
```

File: src/megatron/bridge/training/tensor_inspect.py [hunk 6]
```diff
@@ -0,0 +1,151 @@
...
+
+                def log_scalar(self, name: str, value: float | int, iteration: int, **kwargs):  # type: ignore[override]
+                    self._wandb.log({name: value}, step=iteration)
+
+            MetricLogger.add_logger(_WandbModuleLogger(wandb_logger))
+    except Exception as e:  
+        print_rank_0(f"Skipping NVIDIA DLFw Inspect metric logger attach due to error: {e}")
+
+
+def finalize_tensor_inspect_post_model(
```
Comment (ananthsub) [inline src/megatron/bridge/training/tensor_inspect.py]: the naming is a little ambiguous. post_model_initialization?

File: src/megatron/bridge/training/tensor_inspect.py [hunk 7]
```diff
@@ -0,0 +1,151 @@
...
+    tensor_inspect_config: TensorInspectConfig | None,
+    model: Any,
```
Comment (ananthsub) [inline src/megatron/bridge/training/tensor_inspect.py]:
```suggestion
    model: list[MegatronModule],
```

**Comment by paul-gibbons** [comment]
> @paul-gibbons structurally this looks good to me. could you share what runs with this enabled look like? anything we can add to the docs for this feature? is NVIDIA-DLFW-Inspect considered API stable?

I will add more references to the [docs available in TE ](https://github.com/NVIDIA/TransformerEngine/tree/main/docs/debug) and work on more examples in this MR. 

I will let Anmol and Shreyas comment on any plans for changing DLFW-Inspect API. The integration in TE would be where the majority of changes are happening in order to support collection of new statistics and modules in TE and user's would be expected to bump their TE version more frequently to make use of newer stats.

Here is an example of what a run would look like with this enabled with the following config. Here we are logging what percentage of gradients of all FC1 & FC1 modules in our model underflow to zero, large degree of underflow can indicate divergence during training. 

```
        cfg.tensor_inspect = TensorInspectConfig(
            enabled=True,
            features="/pg/dev/code/nemo_scripts/bridge_configs/fp8_inspect.yaml",
            log_dir="/pg/dev/code/logs",
        )
```

**fp8_inspect.yaml (can also pass feature as python dict)**
```
fp8_tensor_stat_collection:
    enabled: true
    layers:
        layer_name_regex_pattern: ".*(fc1|fc2)"
    transformer_engine:
        LogFp8TensorStats:
            enabled: True
            tensors:
                - gradient
            stats:
                - "underflows%"
            freq: 10
            start_step: 1
            end_step: 10000
```

Screenshot 2025-09-23 at 12 20 38 PM

You can expect overhead at the frequency in which you collect your statistics but there is caching mechanism in NV-DLFW-Inspect and it's integration in TE to help minimize overall overhead when the tooling is in use. 

Screenshot 2025-09-23 at 12 29 52 PM

File: src/megatron/bridge/training/setup.py [hunk 8]
```diff
@@ -161,6 +165,11 @@ def setup(
     timers("tokenizer-setup").stop()
     barrier_and_log("after tokenizer is built")
 
+    # Initialize NVIDIA DLFw Inspect early (this must happen before TE modules are constructed)
+    if cfg.tensor_inspect is not None:
+        cfg.tensor_inspect.finalize()
```
Comment (ananthsub) [inline src/megatron/bridge/training/setup.py]: nit: you could add this in https://github.com/NVIDIA-NeMo/Megatron-Bridge/blob/5f7d239d734c086f619b38fddd9d8e6aca3a67e3/src/megatron/bridge/training/config.py#L965

File: src/megatron/bridge/utils/import_utils.py [hunk 9]
```diff
@@ -42,6 +42,9 @@
     "nvidia-resiliency-ext is not available. Please install it with `pip install nvidia-resiliency-ext`."
 )
 MISSING_NEMO_RUN_MSG = "nemo-run is not available. Please install it with `pip install nemo-run`."
+MISSING_NVINSPECT_MSG = (
```
Comment (ananthsub) [inline src/megatron/bridge/utils/import_utils.py]: i don't expect other files to need to import this considering how self-contained the integration is. this could go directly in tensor_inspect.py

File: docs/training/logging.md [hunk 10]
```diff
@@ -167,6 +167,132 @@ At each checkpoint boundary, the log is updated with:
...
+    LogFp8TensorStats:
+      enabled: true
+      tensors: [weight,activation,gradient]
+      stats: ["underflows%", "mse"]
+      freq: 5
+      start_step: 0
+      end_step: 100
+```
+
+#### Python Configuration
```
Comment (ananthsub) [inline docs/training/logging.md:239]: nit: list the python config over yaml

File: src/megatron/bridge/training/config.py [hunk 11]
```diff
@@ -863,6 +863,45 @@ def finalize(self) -> None:
...
+    """Enable tensor inspection and statistics collection."""
+
+    features: Optional[Union[dict[str, Any], str, Path]] = None
+    """Feature configuration as a Python dict or a YAML file path."""
+
+    feature_dirs: Optional[list[str]] = None
+    """Directories containing feature implementations (searched recursively)."""
+
+    log_dir: Optional[str] = None
+    """Root directory to store inspection logs/statistics. Defaults to checkpoint save dir if unset."""
```
Comment (ananthsub) [inline src/megatron/bridge/training/config.py]:
```suggestion
    features: dict[str, Any] | str, Path | None = None
    """Feature configuration as a Python dict or a YAML file path."""

    feature_dirs: list[str] | None = None
    """Directories containing feature implementations (searched recursively)."""

    log_dir: str | None = None
    """Root directory to store inspection logs/statistics. Defaults to checkpoint save dir if unset."""
```

File: src/megatron/bridge/training/config.py [hunk 12]
```diff
@@ -1047,6 +1086,7 @@ class ConfigContainer(Container):
     peft: Optional[PEFT] = None
     comm_overlap: Optional[CommOverlapConfig] = None
     mixed_precision: Optional[Union[MixedPrecisionConfig, str]] = None
+    tensor_inspect: Optional[TensorInspectConfig] = None
```
Comment (ananthsub) [inline src/megatron/bridge/training/config.py]:
```suggestion
    tensor_inspect: TensorInspectConfig | None = None
```

File: src/megatron/bridge/training/tensor_inspect.py [hunk 13]
```diff
@@ -0,0 +1,153 @@
...
+            default_logging_enabled=True,
+        )
+        print_rank_0("Initialized NVIDIA-DL-Framework-Inspect (pre-model).")
+    except Exception as e:
+        # Treat initialization failures as fatal when enabled so training exits
+        print_rank_0(f"NVIDIA DLFw Inspect pre-init failed: {e}")
+        raise
+
+
+def _maybe_attach_metric_loggers(tensorboard_logger: Optional[Any], wandb_logger: Optional[Any]) -> None:
```
Comment (ananthsub) [inline src/megatron/bridge/training/tensor_inspect.py]:
```suggestion
def _maybe_attach_metric_loggers(tensorboard_logger: Any | None, wandb_logger: Any | None) -> None:
```

File: src/megatron/bridge/training/tensor_inspect.py [hunk 14]
```diff
@@ -0,0 +1,153 @@
...
+    except Exception as e:
+        print_rank_0(f"Skipping NVIDIA DLFw Inspect metric logger attach due to error: {e}")
+
+
+def finalize_tensor_inspect_post_model_initialization(
+    tensor_inspect_config: TensorInspectConfig | None,
+    model: list[MegatronModule],
+    tensorboard_logger: Optional[Any],
+    wandb_logger: Optional[Any],
+    current_training_step: Optional[int] = None,
```
Comment (ananthsub) [inline src/megatron/bridge/training/tensor_inspect.py]:
```suggestion
    tensorboard_logger: Any | None,
    wandb_logger: Any | None,
    current_training_step: int | None = None,
```

**Comment by paul-gibbons** [comment]
/ok to test [5549de9](https://github.com/NVIDIA-NeMo/Megatron-Bridge/pull/649/commits/5549de943607f135c387561689d9f1ec42c3de1d)

**Comment by maanug-nv** [comment]
Closes #1116

**Comment by paul-gibbons** [comment]
/ok to test [608f584](https://github.com/NVIDIA-NeMo/Megatron-Bridge/commit/608f58423861084622f55e81703027f25ad23799)

**Comment by paul-gibbons** [comment]
/ok to test [9aef802](https://github.com/NVIDIA-NeMo/Megatron-Bridge/pull/649/commits/9aef802edc17207e8e42111ccd9e8e5c9ba57f0f)

**Comment by paul-gibbons** [comment]
/ok to test [f3d5dc4](https://github.com/NVIDIA-NeMo/Megatron-Bridge/pull/649/commits/f3d5dc47738989ec76cdcbd0436f4569f731cae2)

**Comment by paul-gibbons** [comment]
/ok to test [a94880f](https://github.com/NVIDIA-NeMo/Megatron-Bridge/pull/649/commits/a94880fc193fcd6d158278527e031f6b4b1794d9)

**Comment by paul-gibbons** [comment]
/ok to test [ad1e9cf](https://github.com/NVIDIA-NeMo/Megatron-Bridge/pull/649/commits/ad1e9cf0bb484eca6596fb3e096287d2b81a4de3)

**Comment by paul-gibbons** [comment]
/ok to test [0b84819](https://github.com/NVIDIA-NeMo/Megatron-Bridge/pull/649/commits/0b84819d0714452d42b2b7603e76118c08183c01)

**Comment by paul-gibbons** [comment]
/ok to test [0c8dc9a](https://github.com/NVIDIA-NeMo/Megatron-Bridge/pull/649/commits/0c8dc9a8e5c21ca07bd72d0d8925c2f2a4f4da6b)
